RavenDB-6931 assert @attachments not modified

talweiss1982 opened the pull request.

Commit:
RavenDB-6931 assert @attachments not modified

talweiss1982 commented:
SlowTests.Authentication.AuthenticationLetsEncryptTests.CanGetLetsEncryptCertificateAndRenewIt failure seems unrelated

ayende commented on src/Raven.Server/Documents/DocumentCompare.cs:
```diff
@@ -69,19 +76,30 @@ private static DocumentCompareResult IsMetadataEqualTo(BlittableJsonReaderObject
+
+        private static void ThrowAttachmentsModificationsDetected(string documentId)
+        {
+            throw new InvalidOperationException("Illegal modifications of '@attachmetns' detected for document: " + documentId);
```
```suggestion
            throw new InvalidOperationException("Illegal modifications of '@attachments' detected for document: " + documentId);
```

ayende commented on src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs:
```diff
@@ -182,7 +182,12 @@ protected PatchResult ExecuteOnDocument(DocumentsOperationContext context, strin
                         }
-                        else if (DocumentCompare.IsEqualTo(originalDoc, modifiedDocument, tryMergeMetadataConflicts: true) != DocumentCompareResult.Equal)
+                        else if (DocumentCompare.IsEqualTo(originalDoc, modifiedDocument, 
+                                    new DocumentCompare.DocumentCompareOptions
```
Let's make this into a constant, like the `Default` field there

talweiss1982 replied:
done

ayende commented on src/Raven.Server/Documents/Replication/ConflictManager.cs:
```diff
@@ -199,7 +199,7 @@ private static void InvalidConflictWhenThereIsNone(string id)
             if (existingDoc != null)
             {
-                var compareResult = DocumentCompare.IsEqualTo(existingDoc.Data, incomingDoc, true);
+                var compareResult = DocumentCompare.IsEqualTo(existingDoc.Data, incomingDoc, new DocumentCompare.DocumentCompareOptions{TryMergeMetadataConflicts = true});
```
Let's make this into a constant, too

talweiss1982 replied:
done

ayende commented on src/Raven.Server/Documents/DocumentCompare.cs:
```diff
@@ -11,8 +11,15 @@ namespace Raven.Server.Documents
+        {
+            public bool TryMergeMetadataConflicts;
+            public bool ThrowOnAttachmentModifications;
+            public string DocumentId;
```
Am I missing something, or is this never set?

talweiss1982 replied:
I forgot to set it but if we will use this than we can't use constant options so its allocation vs better erroring, i think it is best to catch the exception on patch only and add a better message there

ayende replied:
Sounds good, yes

ayende committed:
Update src/Raven.Server/Documents/DocumentCompare.cs

Co-Authored-By: talweiss1982 <tal@ayende.com>

Commit:
RavenDB-6931 PR Comments

Commit:
RavenDB-6931 removed document id from document compare options now catching the exception and adding the id so we could avoid allocation of the option

ayende commented on src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs:
```diff
@@ -182,17 +182,30 @@ protected PatchResult ExecuteOnDocument(DocumentsOperationContext context, strin
                             {
-                                putResult = _database.DocumentsStorage.Put(context, originalDocument.Id,
-                                    originalDocument.ChangeVector, modifiedDocument, null, null, originalDocument.Flags);
+                                throw new InvalidOperationException("Illegal modifications of '@attachments' detected for document: " + id);
```
Better to just state that there was an error processing this and use the `ioe` as inner exception

ppekrol replied:
Done

ppekrol committed:
RavenDB-6931
- switched DocumentCompareOptions to readonly struct
- better exception message
